[Navigation] Initial implementation of NavigationActivation

TingPing opened the pull request:
#### 484ef912355d668f088f6a521ff44de759268d2a
<pre>
[Navigation] Initial implementation of NavigationActivation
<a href="https://bugs.webkit.org/show_bug.cgi?id=275711">https://bugs.webkit.org/show_bug.cgi?id=275711</a>

Reviewed by Alex Christensen.

<a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-activation">https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-activation</a>

An activation occurs on any cross-document navigation, this sets
an object that exposes where the navigation came from.

This is not yet complete as it doesn&apos;t handle reactivation,
which happens when navigating to a cached page. Unfortunately
the WPT tests for this are skipped as the cache is disabled
so this will come later.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-push-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-traverse-not-in-entries-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::reload):
(WebCore::FrameLoader::loadDifferentDocumentItem):
(WebCore::FrameLoader::updateNavigationAPIEntries):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/NavigationAction.h:
(WebCore::NavigationAction::navigationAPIType const):
(WebCore::NavigationAction::setNavigationAPIType):

These changes were to fix location.replace() which was exposed
by these tests.

* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledLocationChange::ScheduledLocationChange):
(WebCore::NavigationScheduler::scheduleLocationChange):

* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeEntries):
(WebCore::Navigation::updateForActivation):
(WebCore::Navigation::updateForReactivation):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Navigation.idl:
* Source/WebCore/page/NavigationActivation.cpp: Added.
(WebCore::NavigationActivation::NavigationActivation):
* Source/WebCore/page/NavigationActivation.h: Added.
* Source/WebCore/page/NavigationActivation.idl: Added.

Canonical link: <a href="https://commits.webkit.org/280355@main">https://commits.webkit.org/280355@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d3b07e629fbf722d5ca32bedc5d910726250e33

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56363 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35689 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8835 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/59970 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6799 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58489 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43311 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6993 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45653 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4747 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58392 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33566 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48634 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26517 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30346 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5964 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/5803 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52337 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6236 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/61654 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/271 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6357 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52916 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/271 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48700 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/52799 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12483 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/245 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31516 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32602 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33685 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32349 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/page/NavigationActivation.h:
```diff
@@ -0,0 +1,55 @@
+
+    NavigationNavigationType m_navigationType;
+    Ref<NavigationHistoryEntry> m_entry;
+    RefPtr<NavigationHistoryEntry> m_from;
```
We should probably name this m_fromEntry to be compatible with the ctor parameter name.

rwlbuis replied:
You dd thumbs up but I think did not implemented it yet :)

TingPing replied:
I did just now, I hit resolve when I do that :)

rwlbuis commented on Source/WebCore/loader/DocumentWriter.cpp:
```diff
@@ -241,7 +241,7 @@ bool DocumentWriter::begin(const URL& urlReference, bool dispatch, Document* own
         document->checkedContentSecurityPolicy()->setInsecureNavigationRequestsToUpgrade(existingDocument->checkedContentSecurityPolicy()->takeNavigationRequestsToUpgrade());
 
-    frameLoader->didBeginDocument(dispatch);
+    frameLoader->didBeginDocument(dispatch, triggeringAction ? triggeringAction->navigationAPIType() : std::nullopt);
```
I don't like this part. Can FrameLoader::didBeginDocument request this information? The less files we change the better.

rwlbuis reviewed (commented):
LGTM. Maybe the commit log can have relevant spec link(s).

achristensen07 commented on Source/WebCore/page/NavigationActivation.h:
```diff
@@ -0,0 +1,55 @@
+class NavigationActivation final : public RefCounted<NavigationActivation>, public ScriptWrappable {
+    WTF_MAKE_ISO_ALLOCATED(NavigationActivation);
+public:
+    static Ref<NavigationActivation> create(NavigationNavigationType type, NavigationHistoryEntry& entry, RefPtr<NavigationHistoryEntry>&& fromEntry)
```
nit: this interface is a little inconsistent.  entry should probably be a Ref<NavigationHistoryEntry>&&

TingPing committed:
[Navigation] Initial implementation of NavigationActivation
https://bugs.webkit.org/show_bug.cgi?id=275711

Reviewed by Alex Christensen.

https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigation-activation

An activation occurs on any cross-document navigation, this sets
an object that exposes where the navigation came from.

This is not yet complete as it doesn't handle reactivation,
which happens when navigating to a cached page. Unfortunately
the WPT tests for this are skipped as the cache is disabled
so this will come later.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-initial-about-blank-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-push-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-reload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-cross-origin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-traverse-not-in-entries-expected.txt:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::didBeginDocument):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::reload):
(WebCore::FrameLoader::loadDifferentDocumentItem):
(WebCore::FrameLoader::updateNavigationAPIEntries):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/NavigationAction.h:
(WebCore::NavigationAction::navigationAPIType const):
(WebCore::NavigationAction::setNavigationAPIType):

These changes were to fix location.replace() which was exposed
by these tests.

* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledLocationChange::ScheduledLocationChange):
(WebCore::NavigationScheduler::scheduleLocationChange):

* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::initializeEntries):
(WebCore::Navigation::updateForActivation):
(WebCore::Navigation::updateForReactivation):
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/Navigation.idl:
* Source/WebCore/page/NavigationActivation.cpp: Added.
(WebCore::NavigationActivation::NavigationActivation):
* Source/WebCore/page/NavigationActivation.h: Added.
* Source/WebCore/page/NavigationActivation.idl: Added.

Canonical link: https://commits.webkit.org/280355@main
